File: ts/src/platform-token-manager.test.ts
import { test } from "node:test";
import { strict as assert } from "node:assert";
import { PlatformTokenManager } from "./platform-token-manager.js";
import { staticApiKey } from "./credential.js";
import { NOOP_LOGGER } from "./logger.js";
import { RealmError } from "./errors.js";

interface Call {
  url: string;
  method: string;
  authorization: string | null;
  body?: unknown;
}

function mkFetch(handler: (call: Call) => Response | Promise<Response>): { fetch: typeof fetch; calls: Call[] } {
  const calls: Call[] = [];
  const fetch = (async (input: RequestInfo | URL, init?: RequestInit): Promise<Response> => {
    const url = typeof input === "string" ? input : input.toString();
    const headers = new Headers(init?.headers);
    let body: unknown = undefined;
    if (init?.body) {
      try { body = JSON.parse(String(init.body)); } catch { body = init.body; }
    }
    const c: Call = { url, method: init?.method ?? "GET", authorization: headers.get("authorization"), body };
    calls.push(c);
    return handler(c);
  }) as typeof fetch;
  return { fetch, calls };
}

function loginResponse(access: string, refresh = "rtok-platform", expires = 300): Response {
  return new Response(JSON.stringify({
    status: "ok",
    subject_type: "platform",
    refresh_token: refresh,
    access_token: access,
    expires_in: expires,
  }), { status: 200, headers: { "content-type": "application/json" } });
}

test("session: first call hits /auth/login, subsequent calls within TTL reuse cache", async () => {
  let logins = 0, refreshes = 0;
  const { fetch, calls } = mkFetch((c) => {
    if (c.url.endsWith("/auth/login")) { logins++; return loginResponse(`at_${logins}`); }
    if (c.url.endsWith("/auth/token")) { refreshes++; return loginResponse(`at_r_${refreshes}`); }
    return new Response("nope", { status: 404 });
  });
  let now = 1_700_000_000_000;
  const mgr = new PlatformTokenManager({
    credential: staticApiKey("rk_live_x"), baseUrl: "https://auth.test", fetch, logger: NOOP_LOGGER, now: () => now,
  });

  const t1 = await mgr.getToken();
  const t2 = await mgr.getToken();
  now += 60_000; // well within 300s TTL
  const t3 = await mgr.getToken();
  assert.equal(t1, "at_1");
  assert.equal(t2, "at_1");
  assert.equal(t3, "at_1");
  assert.equal(logins, 1);
  assert.equal(refreshes, 0);
  assert.equal(calls[0]!.method, "POST");
  assert.equal(calls[0]!.url.endsWith("/auth/login"), true);
  // raw api-key only travels in the body, never as a bearer
  assert.equal(calls[0]!.authorization, null);
  assert.equal((calls[0]!.body as { grant_type: string }).grant_type, "platform_api_key");
});

test("session: refreshes via /auth/token when within 30s of expiry", async () => {
  let logins = 0, refreshes = 0;
  const { fetch } = mkFetch((c) => {
    if (c.url.endsWith("/auth/login")) { logins++; return loginResponse(`at_${logins}`, "rtok-1", 60); }
    if (c.url.endsWith("/auth/token")) { refreshes++; return loginResponse(`at_r_${refreshes}`, "rtok-1", 60); }
    return new Response("nope", { status: 404 });
  });
  let now = 1_700_000_000_000;
  const mgr = new PlatformTokenManager({
    credential: staticApiKey("rk_live_x"), baseUrl: "https://auth.test", fetch, logger: NOOP_LOGGER, now: () => now,
  });

  const t1 = await mgr.getToken();
  assert.equal(t1, "at_1");
  now += 25_000; // 35s remaining > 30s skew → reuse
  const t2 = await mgr.getToken();
  assert.equal(t2, "at_1");
  assert.equal(refreshes, 0);
  now += 10_000; // 25s remaining < 30s skew → /auth/token
  const t3 = await mgr.getToken();
  assert.equal(t3, "at_r_1");
  assert.equal(logins, 1);
  assert.equal(refreshes, 1);
});

test("session: /auth/token 401 falls back to /auth/login", async () => {
  let logins = 0;
  const { fetch } = mkFetch((c) => {
    if (c.url.endsWith("/auth/login")) { logins++; return loginResponse(`at_${logins}`, "rtok-1", 60); }
    if (c.url.endsWith("/auth/token")) {
      return new Response(JSON.stringify({ error: { code: "unauthorized", message: "refresh revoked" } }),
        { status: 401, headers: { "content-type": "application/json" } });
    }
    return new Response("nope", { status: 404 });
  });
  let now = 1_700_000_000_000;
  const mgr = new PlatformTokenManager({
    credential: staticApiKey("rk_live_x"), baseUrl: "https://auth.test", fetch, logger: NOOP_LOGGER, now: () => now,
  });
  await mgr.getToken();
  now += 50_000; // force refresh window
  const t = await mgr.getToken();
  assert.equal(t, "at_2"); // re-logged in
  assert.equal(logins, 2);
});

test("session: 401 from /auth/login surfaces RealmError unauthorized", async () => {
  const { fetch } = mkFetch(() => new Response(JSON.stringify({
    error: { code: "unauthorized", message: "bad api key" },
  }), { status: 401, headers: { "content-type": "application/json" } }));
  const mgr = new PlatformTokenManager({
    credential: staticApiKey("rk_live_bad"), baseUrl: "https://auth.test", fetch, logger: NOOP_LOGGER,
  });
  await assert.rejects(() => mgr.getToken(), (e: Error) => {
    return e instanceof RealmError && e.code === "unauthorized";
  });
});

test("session: malformed response surfaces server_error", async () => {
  const { fetch } = mkFetch(() => new Response(JSON.stringify({ wrong: "shape" }), {
    status: 200, headers: { "content-type": "application/json" },
  }));
  const mgr = new PlatformTokenManager({
    credential: staticApiKey("rk_live_x"), baseUrl: "https://auth.test", fetch, logger: NOOP_LOGGER,
  });
  await assert.rejects(() => mgr.getToken(), (e: Error) => {
    return e instanceof RealmError && e.code === "server_error";
  });
});

test("session: never logs the raw api key (redaction smoke test)", async () => {
  const captured: Array<{ msg: string; meta?: Record<string, unknown> }> = [];
  const logger = {
    debug() {}, warn() {}, error() {},
    info(msg: string, meta?: Record<string, unknown>) { captured.push({ msg, meta }); },
  };
  const { fetch } = mkFetch(() => loginResponse("at_xxxx"));
  const mgr = new PlatformTokenManager({
    credential: staticApiKey("rk_live_supersecret_full_value"),
    baseUrl: "https://auth.test", fetch, logger,
  });
  await mgr.getToken();
  for (const ev of captured) {
    const blob = JSON.stringify(ev);
    assert.ok(!blob.includes("supersecret_full_value"), `raw API key leaked into log: ${blob}`);
    assert.ok(!blob.includes("at_xxxx"), `raw access token leaked into log: ${blob}`);
  }
  assert.ok(captured.length >= 1, "expected at least one info-level log event");
});

test("session: token-exchange credential posts grant_type + subject_token (ADR-057)", async () => {
  const { fetch, calls } = mkFetch((c) => {
    if (c.url.endsWith("/auth/login")) return loginResponse("at_fed");
    return new Response("nope", { status: 404 });
  });
  const cred = {
    async fetch() {
      return {
        grantType: "urn:ietf:params:oauth:grant-type:token-exchange",
        subjectToken: "workload.jwt.tok",
      };
    },
  };
  const mgr = new PlatformTokenManager({
    credential: cred, baseUrl: "https://auth.test", fetch, logger: NOOP_LOGGER,
  });
  const t = await mgr.getToken();
  assert.equal(t, "at_fed");
  const body = calls[0]!.body as Record<string, string>;
  assert.equal(body.grant_type, "urn:ietf:params:oauth:grant-type:token-exchange");
  assert.equal(body.subject_token, "workload.jwt.tok");
  assert.equal(body.subject_token_type, "urn:ietf:params:oauth:token-type:jwt");
  assert.equal(body.api_key, undefined);
});

File: java/src/test/java/dev/realmid/sdk/platformtoken/PlatformTokenManagerTest.java
package dev.realmid.sdk.platformtoken;

import com.fasterxml.jackson.databind.ObjectMapper;
import dev.realmid.sdk.ErrorCode;
import dev.realmid.sdk.FakeServer;
import dev.realmid.sdk.RealmException;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.net.http.HttpClient;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.time.ZoneId;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;

/**
 * Two-endpoint platform session (SPEC §4.0, ADR-051). First acquire goes
 * through {@code POST /auth/login {grant_type:"platform_api_key"}}; subsequent
 * refreshes go through {@code POST /auth/token} with the refresh token as the
 * bearer.
 */
class PlatformTokenManagerTest {
    private FakeServer fs;

    @BeforeEach void setUp() throws IOException { fs = new FakeServer(); }
    @AfterEach void tearDown() { fs.close(); }

    private PlatformTokenManager manager(Clock clock) {
        return new PlatformTokenManager(CredentialSources.staticApiKey("rk_live_secret"), fs.baseUrl,
                HttpClient.newHttpClient(), new ObjectMapper(), null,
                clock, Duration.ofSeconds(30));
    }

    private static Clock mutableClock(long[] nowMs) {
        return new Clock() {
            @Override public ZoneId getZone() { return ZoneId.of("UTC"); }
            @Override public Clock withZone(ZoneId z) { return this; }
            @Override public Instant instant() { return Instant.ofEpochMilli(nowMs[0]); }
            @Override public long millis() { return nowMs[0]; }
        };
    }

    @Test
    void loginAndCacheHit() {
        AtomicInteger calls = new AtomicInteger();
        fs.on("POST /auth/login", (ex, body) -> {
            calls.incrementAndGet();
            return FakeServer.Reply.json(200, Map.of(
                    "access_token", "pt-" + calls.get(),
                    "refresh_token", "rt-" + calls.get(),
                    "expires_in", 300,
                    "subject_type", "platform"));
        });
        var ptm = manager(Clock.systemUTC());

        String t1 = ptm.getToken();
        String t2 = ptm.getToken();
        assertEquals(1, calls.get());
        assertEquals("pt-1", t1);
        assertSame(t1, t2);
    }

    @Test
    void remintOnExpiryUsesLogin() {
        // ADR-089: a credential-bootstrapped session has no refresh token, so a
        // stale access token is replaced by re-presenting the credential, not by
        // /auth/token. The login reply deliberately carries NO refresh_token —
        // the shape a v0.68.0 issuer actually returns.
        AtomicInteger logins = new AtomicInteger();
        AtomicInteger refreshes = new AtomicInteger();
        fs.on("POST /auth/login", (ex, body) -> {
            int n = logins.incrementAndGet();
            return FakeServer.Reply.json(200, Map.of(
                    "access_token", "pt-login-" + n,
                    "expires_in", 60, "subject_type", "platform"));
        });
        fs.on("POST /auth/token", (ex, body) -> {
            refreshes.incrementAndGet();
            return FakeServer.Reply.json(200, Map.of(
                    "access_token", "pt-refresh", "expires_in", 60, "subject_type", "platform"));
        });
        long[] nowMs = { Instant.parse("2024-01-01T00:00:00Z").toEpochMilli() };
        var ptm = manager(mutableClock(nowMs));
        String first = ptm.getToken();
        // Advance past skew so the cached access token is stale.
        nowMs[0] += 50_000;
        String second = ptm.getToken();
        assertEquals(2, logins.get(), "a stale token re-mints via /auth/login");
        assertEquals(0, refreshes.get(), "ADR-089: /auth/token must not be called");
        assertNotEquals(first, second);
        assertEquals("pt-login-1", first);
        assertEquals("pt-login-2", second);
    }

    @Test
    void invalidateRemintsFromCredential() {
        AtomicInteger logins = new AtomicInteger();
        fs.on("POST /auth/login", (ex, body) -> {
            int n = logins.incrementAndGet();
            return FakeServer.Reply.json(200, Map.of(
                    "access_token", "pt-login-" + n, "expires_in", 300, "subject_type", "platform"));
        });
        fs.on("POST /auth/token", (ex, body) -> {
            throw new AssertionError("ADR-089: /auth/token must not be called");
        });
        var ptm = manager(Clock.systemUTC());
        assertEquals("pt-login-1", ptm.getToken());
        ptm.invalidate();
        assertEquals("pt-login-2", ptm.getToken());
        assertEquals(2, logins.get());
    }

    @Test
    void ignoresAStrayRefreshTokenInTheLoginReply() {
        // Interop with a PRE-ADR-089 issuer, which still returns a refresh
        // token: the manager must ignore it rather than start using
        // /auth/token, which on a current issuer answers
        // 401 m2m_refresh_withdrawn.
        AtomicInteger logins = new AtomicInteger();
        fs.on("POST /auth/login", (ex, body) -> {
            int n = logins.incrementAndGet();
            return FakeServer.Reply.json(200, Map.of(
                    "access_token", "pt-login-" + n, "refresh_token", "rt-" + n,
                    "expires_in", 300, "subject_type", "platform"));
        });
        fs.on("POST /auth/token", (ex, body) -> FakeServer.Reply.json(401,
                Map.of("error", Map.of("code", "m2m_refresh_withdrawn", "message", "no refresh token"))));
        var ptm = manager(Clock.systemUTC());
        ptm.getToken();
        ptm.invalidate();
        String tok = ptm.getToken();
        assertEquals(2, logins.get(), "the stray refresh token must not be used");
        assertEquals("pt-login-2", tok);
    }

    @Test
    void unauthorizedLoginSurfaces() {
        fs.on("POST /auth/login", (ex, body) -> FakeServer.Reply.json(401,
                Map.of("error", Map.of("code", "unauthorized", "message", "bad key"))));
        var ptm = manager(Clock.systemUTC());
        RealmException ex = assertThrows(RealmException.class, ptm::getToken);
        assertEquals(ErrorCode.UNAUTHORIZED, ex.getCode());
        assertEquals(401, ex.getHttpStatus());
    }

    @Test
    void tokenExchangeCredentialPostsSubjectToken() {
        fs.on("POST /auth/login", (ex, body) -> FakeServer.Reply.json(200, Map.of(
                "access_token", "pt-fed", "refresh_token", "rt-fed",
                "expires_in", 300, "subject_type", "platform")));
        CredentialSource cred = () -> Credential.ofWorkloadToken("workload.jwt.tok");
        var ptm = new PlatformTokenManager(cred, fs.baseUrl, HttpClient.newHttpClient(),
                new ObjectMapper(), null, Clock.systemUTC(), Duration.ofSeconds(30));

        assertEquals("pt-fed", ptm.getToken());
        Map<String, Object> body = fs.last().bodyAsMap();
        assertEquals("urn:ietf:params:oauth:grant-type:token-exchange", body.get("grant_type"));
        assertEquals("workload.jwt.tok", body.get("subject_token"));
        assertEquals("urn:ietf:params:oauth:token-type:jwt", body.get("subject_token_type"));
        assertNull(body.get("api_key"));
    }

    // ---- ADR-041 client-side realm pin (SPEC §4.0) ----
    //
    // Go (go/platform_token.go checkIssuer) and TS
    // (ts/src/platform-token-manager.ts checkIssuer) decode the freshly minted
    // platform access token and refuse it when its `iss` does not reference the
    // configured realm. Java carried only ErrorCode.REALM_MISMATCH, so the
    // confused-deputy case — SDK constructed for realm A, credential belonging
    // to realm B — surfaced as a cryptic 4xx on some later management call, or
    // not at all.

    private static final String REALM_A = "01HREALMA";
    private static final String REALM_B = "01HREALMB";

    private PlatformTokenManager managerFor(String realmId) {
        return new PlatformTokenManager(CredentialSources.staticApiKey("rk_live_secret"), fs.baseUrl,
                HttpClient.newHttpClient(), new ObjectMapper(), null,
                Clock.systemUTC(), Duration.ofSeconds(30), realmId);
    }

    /** An unsigned-but-well-formed JWT carrying just `iss`. The pin decodes the
     *  payload without verifying the signature — it just received the token
     *  from RI over TLS; signature checking stays the Verifier's job. */
    private static String jwtWithIssuer(String iss) {
        try {
            String hdr = b64(FakeServer.M.writeValueAsBytes(Map.of("alg", "RS256", "typ", "JWT")));
            Map<String, Object> claims = new LinkedHashMap<>();
            claims.put("iss", iss);
            claims.put("sub", "bot-user");
            String pl = b64(FakeServer.M.writeValueAsBytes(claims));
            return hdr + "." + pl + "." + b64("sig-not-checked-here".getBytes());
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static String b64(byte[] b) {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(b);
    }

    private void serveLoginWithIssuer(String iss, AtomicInteger logins) {
        fs.on("POST /auth/login", (ex, body) -> {
            logins.incrementAndGet();
            return FakeServer.Reply.json(200, Map.of(
                    "access_token", jwtWithIssuer(iss),
                    "expires_in", 300, "subject_type", "platform"));
        });
    }

    @Test
    void realmPinRefusesATokenMintedForAnotherRealm() {
        AtomicInteger logins = new AtomicInteger();
        serveLoginWithIssuer("https://auth.realmid.dev/" + REALM_B, logins);
        var ptm = managerFor(REALM_A);

        RealmException ex = assertThrows(RealmException.class, ptm::getToken);
        assertEquals(ErrorCode.REALM_MISMATCH, ex.getCode());
        assertEquals(1, logins.get());
        // The refused token must not be cached: a second call re-mints and
        // refuses again rather than handing out the foreign-realm token.
        assertThrows(RealmException.class, ptm::getToken);
        assertEquals(2, logins.get());
    }

    @Test
    void realmPinAcceptsATokenMintedForTheConfiguredRealm() {
        // POSITIVE CONTROL. Without it, a pin that refused every token — or one
        // whose suffix comparison is inverted — would satisfy the test above.
        AtomicInteger logins = new AtomicInteger();
        String iss = "https://auth.realmid.dev/" + REALM_A;
        serveLoginWithIssuer(iss, logins);
        var ptm = managerFor(REALM_A);

        String tok = assertDoesNotThrow(ptm::getToken);
        assertEquals(jwtWithIssuer(iss), tok);
        assertEquals(1, logins.get());
    }

    @Test
    void realmPinIgnoresATokenItCannotDecode() {
        // Parity with Go (peekJWTIssuer error → return nil) and TS (peek returns
        // "" → skip): an opaque or malformed access token is NOT a realm
        // mismatch. The pin is a confused-deputy check, not a token validator —
        // a token whose provenance it cannot read is left to the Verifier.
        AtomicInteger logins = new AtomicInteger();
        fs.on("POST /auth/login", (ex, body) -> {
            logins.incrementAndGet();
            return FakeServer.Reply.json(200, Map.of(
                    "access_token", "pt-opaque", "expires_in", 300, "subject_type", "platform"));
        });
        var ptm = managerFor(REALM_A);
        assertEquals("pt-opaque", assertDoesNotThrow(ptm::getToken));
    }

    @Test
    void realmPinIsSkippedWhenNoRealmIdIsConfigured() {
        // The 7-arg constructor predates the pin and carries no realm to pin
        // against; it must keep working rather than start refusing. Mirrors
        // TS's `if (!this.opts.realmId) return`.
        AtomicInteger logins = new AtomicInteger();
        serveLoginWithIssuer("https://auth.realmid.dev/" + REALM_B, logins);
        var ptm = manager(Clock.systemUTC());
        assertDoesNotThrow(ptm::getToken);
    }
}
